File: clojure/bob/src/bob.clj
(ns bob
  (:require [clojure.string :as s]))

(defn yelling? [s]
  (and (= s (s/upper-case s))
       (some #(Character/isLetter %1) s)))

(defn question? [s]
  (s/ends-with? s "?"))

(defn nothing-said? [s]
  (= "" (s/trim s)))

(defn response-for [s]
  (cond
    (nothing-said? s) "Fine. Be that way!"
    (and (yelling? s) (question? s)) "Calm down, I know what I'm doing!"
    (yelling? s) "Whoa, chill out!"
    (question? s) "Sure."
    :else "Whatever."))

File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Char (isLetter, isSpace, isUpper)

responseFor :: String -> String
responseFor xs
    | isSilent = "Fine. Be that way!"
    | isQuestion && isShouting = "Calm down, I know what I'm doing!"
    | isShouting = "Whoa, chill out!"
    | isQuestion = "Sure."
    | otherwise = "Whatever."
    where text = filter (not . isSpace) xs 
          letters = filter isLetter text
          isQuestion = last text  == '?'
          isShouting = all isUpper letters && any isUpper letters
          isSilent = null text
